### PR TITLE
V5 migration of get-balance (native) and get-all-events api

### PR DESCRIPTION
**Pull request by nischitpra**
Unit tests are created for compairing v4 and v5 outputs.
I'll need some basic help with Typescript types. I've only made jsdoc comments to explain the function parameters.

APIs migrated
- get-balance
- get-all-events

---

## PR-Codex overview
This PR primarily focuses on code quality improvements, including type adjustments, error handling enhancements, and updates to logging messages. It also includes some updates to dependencies and modifications to the way certain functions are structured.

### Detailed summary
- Added `text=auto` to `.gitattributes`.
- Updated error handling to use `_e` instead of `e` in multiple files.
- Changed message formatting in several log statements.
- Adjusted type imports to use `type` keyword where applicable.
- Refactored function parameters to use `unknown` instead of `any`.
- Updated `package.json` version for `@thirdweb-dev/engine`.
- Improved consistency in using `const` instead of `let` where possible.
- Modified handler signatures to use `_req` instead of `req` in various routes.
- Added a new `testWebhook` method in `WebhooksService`.
- Enhanced compatibility for chain ID handling in multiple services.

> The following files were skipped due to too many changes: `sdk/src/services/AccountFactoryService.ts`, `tests/unit/migrationV5.test.ts`, `sdk/src/services/ContractRolesService.ts`, `sdk/src/services/ContractRoyaltiesService.ts`, `sdk/src/services/MarketplaceOffersService.ts`, `sdk/src/services/AccountService.ts`, `sdk/src/services/MarketplaceEnglishAuctionsService.ts`, `sdk/src/services/BackendWalletService.ts`, `sdk/src/services/MarketplaceDirectListingsService.ts`, `sdk/src/services/DeployService.ts`, `sdk/src/services/Erc20Service.ts`, `sdk/src/services/Erc721Service.ts`, `sdk/src/services/Erc1155Service.ts`, `yarn.lock`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: tests/unit/migrationV5.test.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import { getChain } from "../../src/shared/utils/chain";
+import { thirdwebClient } from "../../src/shared/utils/sdk";
+import { getWalletBalance } from "thirdweb/wallets";
+import { getBalance } from "thirdweb/extensions/erc20";
+import { getContractEvents } from "thirdweb";
+import { getContract as getContractV5 } from "thirdweb";
+import { getContract as getContractV4 } from "../../src/shared/utils/cache/get-contract";
+import { mapEventsV4ToV5 } from "../../src/server/routes/contract/events/get-all-events";
+
+/**
```
Comment (arcoraven) [inline tests/unit/migrationV5.test.ts:13]: Hopefully these tests were autogen, thanks for adding them!

Don't need this comment here. The thirdweb secret key should be expected in `.env.test` already.

File: src/server/routes/contract/events/get-all-events.ts [hunk 2]
```diff
@@ -82,20 +85,67 @@ export async function getAllEvents(fastify: FastifyInstance) {
...
+      // backwards compatibility of BigInt(v5) to BigNumber(v4)
+      Object.keys(event.args).forEach((key) => {
+        let value = event.args[key];
+        if (typeof value == "bigint") {
+          value = BigNumber.from(value.toString());
+        }
+        data[key] = value;
+      });
+
+      delete event.eventName;
```
Comment (arcoraven) [inline src/server/routes/contract/events/get-all-events.ts:134]: It's less readable and performant to delete fields. It's more readable to explicitly map fields from one object to another, even if it's long

```
const fooV5 = {
  field1: v4.field1,
  field2: v4.field2,
  field3: BigNumber.from(v4.field3),
  ...
}
```

File: src/server/routes/contract/events/get-all-events.ts [hunk 3]
```diff
@@ -82,20 +85,67 @@ export async function getAllEvents(fastify: FastifyInstance) {
...
+      delete event.args;
+      const transaction = event;
+      transaction.blockNumber = parseInt(transaction.blockNumber);
```
Comment (arcoraven) [inline src/server/routes/contract/events/get-all-events.ts:137]: Nit: This prob does the same thing but `Number(someBigInt)` is more clear to me we're casting one type to another

File: src/server/routes/contract/events/get-all-events.ts [hunk 4]
```diff
@@ -82,20 +85,67 @@ export async function getAllEvents(fastify: FastifyInstance) {
...
+      transaction.event = eventName;
+
+      return {
+        eventName,
+        data,
+        transaction,
+      };
+    })
+    .sort((a, b) => {
```
Comment (arcoraven) [inline src/server/routes/contract/events/get-all-events.ts:146]: We shouldn't need to sort. Esp if we make the transformer for a single function only.

File: src/server/routes/contract/events/get-all-events.ts [hunk 5]
```diff
@@ -82,20 +85,67 @@ export async function getAllEvents(fastify: FastifyInstance) {
...
+
+  return eventsV5
+    .map((event) => {
+      const eventName = event.eventName;
+      const data = {};
+
+      // backwards compatibility of BigInt(v5) to BigNumber(v4)
+      Object.keys(event.args).forEach((key) => {
+        let value = event.args[key];
+        if (typeof value == "bigint") {
```
Comment (arcoraven) [inline src/server/routes/contract/events/get-all-events.ts:128]: Always use `===` in typescript

File: src/server/routes/contract/events/get-all-events.ts [hunk 6]
```diff
@@ -82,20 +85,67 @@ export async function getAllEvents(fastify: FastifyInstance) {
...
+export function mapEventsV4ToV5(eventsV5 = [], order = "desc") {
+  if (!eventsV5?.length) return [];
+
+  return eventsV5
+    .map((event) => {
+      const eventName = event.eventName;
+      const data = {};
+
+      // backwards compatibility of BigInt(v5) to BigNumber(v4)
+      Object.keys(event.args).forEach((key) => {
```
Comment (arcoraven) [inline src/server/routes/contract/events/get-all-events.ts:126]: I think you don't need this if you're explicitly mapping as mentioned below

File: src/server/routes/contract/events/get-all-events.ts (hunk 2 above)
Comment (nischitpra) [reply src/server/routes/contract/events/get-all-events.ts:134]: But if the library updates or the data structure adds new fields, we'll have to remember to add it here. Though typescript will probably give error for that as well and not build so might as well map each fields manually :+1:

File: tests/unit/migrationV5.test.ts (hunk 1 above)
Comment (nischitpra) [reply tests/unit/migrationV5.test.ts:13]: autogen? I write my own code!

File: src/server/routes/contract/events/get-all-events.ts (hunk 6 above)
Comment (nischitpra) [reply src/server/routes/contract/events/get-all-events.ts:126]: Its required as the args object can be anything as discussed in call

File: src/server/routes/contract/events/get-all-events.ts (hunk 4 above)
Comment (nischitpra) [reply src/server/routes/contract/events/get-all-events.ts]: old api has sort functionality so its required for now

**Review by arcoraven — APPROVED** [review]

**Pull request closed without merging**